### PR TITLE
Auto generate event reference docs

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
This follows up on #2698 and autogenerates a reference table of all of the events one can connect to on `Viewer` and `Layer`:

ev

Untitled

I've started (but not completed) a pattern suggested by @goanpeca whereby we can document events in the code itself using `pydantic.Field`:

```python
class Dims(EventedModel):
    ndim: int = Field(default=2, title=trans._("Number of dimensions"))
    ndisplay: Literal[2, 3] = Field(
        default=2, title=trans._("Number of displayed dimensions")
    )
   ...
```

makes the code itself a little uglier ... but keeps all of the "meaning" in one place, which is nice.  however, doing so on the Layer events themselves would require a new construct (since they are not evented models)... but they could be populated manually here for now as well.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by goanpeca** [comment]
>makes the code itself a little uglier ...

🙃 

>(since they are not evented models)

Yet?

Or they will never be 🤔

**Comment by tlambert03** [comment]
yet...  we've long planned to make them evented models.

File: napari/components/dims.py [hunk 1]
```diff
@@ -65,13 +65,33 @@ class Dims(EventedModel):
...
+    ndisplay: Literal[2, 3] = Field(
+        default=2, title=trans._("Number of displayed dimensions")
+    )
+    last_used: int = Field(default=0, title=trans._("Last-used dimension"))
+    range: Tuple[Tuple[float, float, float], ...] = Field(
+        default_factory=tuple,
+        title=trans._("Range in each dimension"),
+        description=trans._(
+            "(min_val, max_val, step_size) for each dimension."
+        ),
```
Comment (sofroniewn) [inline napari/components/dims.py]: >  document events in the code itself using pydantic.Field ....
> makes the code itself a little uglier ... but keeps all of the "meaning" in one place, which is nice.

Is there a way we could parse the existing doc string of `Dims` and then add these, or would that not serve any additional purpose. I do worry that it leads to quite a bit of doc string duplication and makes reading the class a bit harder

**Review by sofroniewn — COMMENTED** [review]
Just started looking at this more. I think having a table like this will be really helpful to folks!

Is the autogeneration approach similar to how we did the autogeneration of preferences docs in #2672? cc @goanpeca. I would probably be nice if they both took as much as the same approach as possible, in terms of how/ when they are run as that could make maintenance a little easier.

> document events in the code itself using pydantic.Field 

I'm a little less sure about this part - I like the vision, but have some concerns about how it looks, as noted in my comment above. Maybe we pull this part out into it's own issue/ PR and we can discuss further there. It's not required for this PR is it?

**Comment by tlambert03** [comment]
> Is the autogeneration approach similar to how we did the autogeneration of preferences docs in #2672? cc @goanpeca. I would probably be nice if they both took as much as the same approach as possible

in the sense that they're both just scripts that drop a new markdown file in the docs folder, yep!


> > document events in the code itself using pydantic.Field

> I'm a little less sure about this part - I like the vision, but have some concerns about how it looks, as noted in my comment above. Maybe we pull this part out into it's own issue/ PR and we can discuss further there. It's not required for this PR is it?

I think the main thing we really want to avoid here is having this info in more than one place.  It's not required for this PR, but without some sort of description, each event is probably going to look a little "bare" (in terms of descriptive usefulness).  The benefit of including it in the `Field` is that it then becomes part of the schema (or the json schema) exported by the model.  If it's mostly the additional line breaks that you don't like, then we could easily have the "one place" be elsewhere, as a dict, and have each Field pull the info as in:
```python
attr_docs = {
    'Dims.order': trans._("For display, the last dimensions are rendered."),
}

class Dims:
    order: Tuple[int, ...] = Field(
        default_factory=tuple,
        description=attr_docs['Dims.order']
    )
```

Of course the last alternative is just to maintain two separate places that info about any given attribute is stored...

**Comment by sofroniewn** [comment]
> I think the main thing we really want to avoid here is having this info in more than one place. It's not required for this PR, but without some sort of description, each event is probably going to look a little "bare" (in terms of descriptive usefulness).

I guess what I was thinking was if we could pull this info straight from the class docstring (where there the parameters/ attributes should be documented anyway). That seems like it would leave us with only one copy of the doc string. Is this a possible approach?

```python
class Dims(EventedModel):
    """Dimensions object modeling slicing and displaying.

    Parameters
    ----------
    ndim : int
        Number of dimensions.
    ndisplay : int
        Number of displayed dimensions.
    last_used : int
        Dimension which was last used.
    range : tuple of 3-tuple of float
        List of tuples (min, max, step), one for each dimension. In a world
        coordinates space.
    current_step : tuple of int
        Tuple the slider position for each dims slider, in slider coordinates.
    order : tuple of int
        Tuple of ordering the dimensions, where the last dimensions are rendered.
    axis_labels : tuple of str
        Tuple of labels for each dimension.
```

**Comment by goanpeca** [comment]
>I guess what I was thinking was if we could pull this info straight from the class docstring (where there the parameters/ attributes should be documented anyway). That seems like it would leave us with only one copy of the doc string. Is this a possible approach?

It would be easier to build a docstring from the `Fields` than "parse" a docstring to get the information needed, no?

**Comment by tlambert03** [comment]
I would still love to have better documentation on events... perhaps the docs working group could take some of the ideas here into consideration.  (This PR offers one possibility, but there are certainly others).  cc @DragaDoncila @melissawm @GenevieveBuckley 

if desired, this can be closed in the PR cleanup on 10/25

**Comment by GenevieveBuckley** [comment]
> I would still love to have better documentation on events... perhaps the docs working group could take some of the ideas here into consideration. (This PR offers one possibility, but there are certainly others). cc @DragaDoncila @melissawm @GenevieveBuckley
> 
> if desired, this can be closed in the PR cleanup on 10/25

I don't have good context on this PR, can you summarize why it's stalled @tlambert03? cc @neuromusic 

What I *think* I understand, from a quick look through is that there are two parts to this PR:
1. Manually created tables in ` docs/guides/events_reference.md`, and
2. Extra code intended to insert event documentation into docstrings.

It seems like 2 is contentious, and Nick has some concerns about how it would go (both the look in docstrings, and the ongoing maintenance required).

Could these two parts be split apart? It doesn't seem as though anyone objects to 1, so perhaps we can include that. Then the discussion around 2 can be resolved separately.

For 1, there are a lot of empty rows in the column `Emitted when __ changes`. I imagine this would need to be fixed before it could be merged.

**Comment by sofroniewn** [comment]
> It seems like 2 is contentious, and Nick has some concerns about how it would go (both the look in docstrings, and the ongoing maintenance required).

Yeah i'm curious if we have any better approaches now? If not and people are ok with the current one then I won't block it, but it did seem like quite a bit of duplication inside the files

> Could these two parts be split apart? It doesn't seem as though anyone objects to 1, so perhaps we can include that. Then the discussion around 2 can be resolved separately.

That would be fine with me!

**Comment by tlambert03** [comment]
>What I think I understand, from a quick look through is that there are two parts to this PR:
> 1. Manually created tables in docs/guides/events_reference.md, and
> 2. Extra code intended to insert event documentation into docstrings.

Hey @GenevieveBuckley, the real core of this PR the [`update_event_docs.py` script](https://github.com/napari/napari/pull/2750/files?authenticity_token=jBjgS%2BA2vp5T9u8sFcVFrEMdxBkLpFE0NIoeISf30EEbO4xHtwPxFimN1nn5mChQppLSWS0wpcKpFU2xdGpOmg%3D%3D&file-filters%5B%5D=.py&file-filters%5B%5D=.yml#diff-9b04aacea383fb7dac58b216c9610562d80761e50ef72dfcdb157196dd4d3de7).  _That_ is what generates the `events_reference.md` table (it wasn't manually created), and the whole goal here is to make it so that no manually editing was needed.

> For 1, there are a lot of empty rows in the column Emitted when __ changes. I imagine this would need to be fixed before it could be merged.

This really was the holdup.  Because it's not that someone needs to go through and manually update the table, we need to a) decide _where_ that content is going to go (currently it doesn't exist anywhere), and then b) have someone go through and make sure that info is in the code.

there are two options:
1. add descriptions of each attribute/event in the `Field()` declared on the model (as done [here](https://github.com/napari/napari/pull/2750/files/d892c88a02138b0e023f9bf78d290b4e5de5039c#diff-ab5508adaadc17d3bbef66e87273ee36aadc1f0ed993bb749c0280851fafd03bR68-R78) in this PR).  The benefit is that it becomes very easy to parse and find this information, and it becomes part of the model "schema" (by nature of being included in `Field`).  the downside is that we then either have potentially duplicated information in the docstrings, or sparse docstrings, etc...  
2.  pull that information from the docstring.  This of course minimizes duplication,  but it becomes a lot harder to parse.  A little extra logic is required to cleanly extract that information from the docstrings (the parsing of which can be error prone).

and that's where the conversation stalled

**Comment by sofroniewn** [comment]
> there are two options:

I can live with option 1. if there is at least one other vote in favor - @goanpeca @jni, want to weigh in?

**Comment by jni** [comment]
Discussion during the PR party came to the consensus of Option 2: use docstrings, with validation during testing to make sure the docstring approach is working.

**Comment by tlambert03** [comment]
I updated the event extraction logic to use the text in the docstrings.  There's a lot of places where it doesn't exactly "fit" great (in terms of making sense in the context of an events table) since it's describing the attribute itself, and not the event of the attribute changing ... but that can be improved in a follow up.

This is all set to pass off...

Screen Shot 2021-11-01 at 4 19 55 PM

**Comment by DragaDoncila** [comment]
> There's a lot of places where it doesn't exactly "fit" great (in terms of making sense in the context of an events table) since it's describing the attribute itself, and not the event of the attribute changing

I don't see this as a critical issue for a couple of reasons:
- events are likely being used by users who are more familiar with napari as a whole, so they're more likely to find the event they want even if the description is not technically a perfect match
- the benefit of automating this extraction far outweighs the cons of having it be slightly less readable
- we have *no* events documentation right now and this is already a huge step forward

How much trouble would it be to also parse where the events are emitted i.e. the function signatures in which x.events.y() is called @tlambert03? I assume we'd have to navigate through the AST so.. a lot? I ask because it would be nice to be able to point users directly to those locations in the codebase.

Alongside these tables maybe we should also have an overview explanation about when common events are emitted, and to point out what events are common to all layers. For example we could mention that in general, layer events are emitted when that particular attribute is directly set, but not necessarily when its contents are edited e.g. `layer.data`. Or maybe there isn't enough commonality among the different types of events for that to be meaningful.

**Comment by tlambert03** [comment]
> How much trouble would it be to also parse where the events are emitted i.e. the function signatures in which x.events.y() is called @tlambert03?

yeah, a lot 😂 ... one is also likely to miss a number of places... since it's tough to resolve dynamic names when traversing AST  (e.g. what event is getting emitted here... `getattr(self.events, name)(value=value)`)

> Alongside these tables maybe we should also have an overview explanation about when common events are emitted, and to point out what events are common to all layers. For example we could mention that in general, layer events are emitted when that particular attribute is directly set, but not necessarily when its contents are edited e.g. layer.data. Or maybe there isn't enough commonality among the different types of events for that to be meaningful.

totally agree.  would be nice to extend this with some manual prose.  I'd appreciate it if that could be done in a followup PR, the automated extraction was the goal here, and is now accomplished

**Comment by DragaDoncila** [comment]
> I'd appreciate it if that could be done in a followup PR, the automated extraction was the goal here, and is now accomplished

Completely agree. This is more than worth going in as is.

**Review by sofroniewn — APPROVED** [review]
This looks great! Very happy to have this go in as a first pass. thanks for the helpful comments @DragaDoncila too. We can leave open to tomorrow if @jni or @GenevieveBuckley have any additional comments to change before merge. Otherwise i think always fine for us to continue to iterate and improve this after merge too, so no worries if you don't have time to look right away.

**Comment by GenevieveBuckley** [comment]
No further comments. This is great to have, thank you @tlambert03!
